File: migalha/Label.java
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

// tracks how information from sources flows and which sanitizers were applied for one vulnerability pattern
public class Label {
    // key - the name of the source
    // value - name of the sanitizers applied to information of that source
    private final Map<String, Set<String>> flows;

    public Label() {
        this.flows = new HashMap<>();
    }

    public Label(Label other) {
        this.flows = new HashMap<>();
        for (Map.Entry<String, Set<String>> entry : other.flows.entrySet()) {
            this.flows.put(entry.getKey(), new HashSet<>(entry.getValue()));
        }
    }

    public void addSource(String source) {
        flows.putIfAbsent(source, new HashSet<>());
    }

    public void addSanitizer(String source, String sanitizer) {
        flows.putIfAbsent(source, new HashSet<>());
        flows.get(source).add(sanitizer);
    }

    public Map<String, Set<String>> getFlows() {
        Map<String, Set<String>> copy = new HashMap<>();
        for (String src : flows.keySet()) {
            copy.put(src, Collections.unmodifiableSet(flows.get(src)));
        }
        return Collections.unmodifiableMap(copy);
    }

    public Set<String> getSources() {
        return Collections.unmodifiableSet(flows.keySet());
    }

    public Set<String> getSanitizersFor(String source) {
        return flows.containsKey(source)
                ? Collections.unmodifiableSet(flows.get(source))
                : Collections.emptySet();
    }

    public Label combine(Label other) {
        Label combined = new Label(this);

        for (Map.Entry<String, Set<String>> entry : other.flows.entrySet()) {
            String src = entry.getKey();
            Set<String> sanitizers = entry.getValue();

            combined.flows.putIfAbsent(src, new HashSet<>());
            combined.flows.get(src).addAll(sanitizers);
        }

        return combined;
    }
}

File: migalha/Label.py
from copy import deepcopy
from typing import Dict, Set, FrozenSet


class Label:
    """
    Tracks how information from sources flows and which sanitizers
    were applied for one vulnerability pattern.
    """

    def __init__(self, other: "Label" = None):
        if other is None:
            # key - the name of the source
            # value - name of the sanitizers applied to information of that source
            self._flows: Dict[str, Set[str]] = {}
        else:
            self._flows = {src: set(sanitizers) for src, sanitizers in other._flows.items()}

    def add_source(self, source: str) -> None:
        self._flows.setdefault(source, set())

    def add_sanitizer(self, source: str, sanitizer: str) -> None:
        self._flows.setdefault(source, set()).add(sanitizer)

    def get_flows(self) -> Dict[str, FrozenSet[str]]:
        # Returns immutable copies using frozenset
        return {src: frozenset(sanitizers) for src, sanitizers in self._flows.items()}

    def get_sources(self) -> FrozenSet[str]:
        return frozenset(self._flows.keys())

    def get_sanitizers_for(self, source: str) -> FrozenSet[str]:
        return frozenset(self._flows.get(source, set()))

    def combine(self, other: "Label") -> "Label":
        combined = Label(self)  # copy of self
        for src, sanitizers in other._flows.items():
            combined._flows.setdefault(src, set()).update(sanitizers)
        return combined

    def __repr__(self):
        return f"Label({self._flows})"
